File: android/src/main/java/com/adyenreactnativesdk/cse/AdyenCSEModule.kt
/*
 * Copyright (c) 2023 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 */

package com.adyenreactnativesdk.cse

import com.adyen.checkout.cse.CardEncrypter
import com.adyen.checkout.cse.EncryptionException
import com.adyen.checkout.cse.UnencryptedCard
import com.adyen.checkout.core.CardBrand
import com.adyen.checkout.core.ui.model.ExpiryDate
import com.adyen.checkout.core.ui.validation.CardExpiryDateValidationResult
import com.adyen.checkout.core.ui.validation.CardExpiryDateValidator
import com.adyen.checkout.core.ui.validation.CardNumberValidationResult
import com.adyen.checkout.core.ui.validation.CardNumberValidator
import com.adyen.checkout.core.ui.validation.CardSecurityCodeValidationResult
import com.adyen.checkout.core.ui.validation.CardSecurityCodeValidator
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableNativeMap

class AdyenCSEModule(
  context: ReactApplicationContext?,
) : ReactContextBaseJavaModule(context) {
  @ReactMethod
  fun addListener(eventName: String?) { // No JS events expected
  }

  @ReactMethod
  fun removeListeners(count: Int?) { // No JS events expected
  }

  override fun getName() = COMPONENT_NAME

  @ReactMethod
  fun encryptCard(
    card: ReadableMap,
    publicKey: String,
    promise: Promise,
  ) {
    val unencryptedCardBuilder = UnencryptedCard.Builder()
    card.getString(NUMBER_KEY)?.let { unencryptedCardBuilder.setNumber(it) }
    val month = card.getString(EXPIRY_MONTH_KEY)
    val year = card.getString(EXPIRY_YEAR_KEY)
    if (month != null && year != null) {
      unencryptedCardBuilder.setExpiryDate(month, year)
    }
    card.getString(CVV_KEY)?.let { unencryptedCardBuilder.setCvc(it) }

    try {
      val encryptedCard =
        CardEncrypter.encryptFields(unencryptedCardBuilder.build(), publicKey)
      val map = WritableNativeMap()
      map.putString(NUMBER_KEY, encryptedCard.encryptedCardNumber)
      map.putString(EXPIRY_MONTH_KEY, encryptedCard.encryptedExpiryMonth)
      map.putString(EXPIRY_YEAR_KEY, encryptedCard.encryptedExpiryYear)
      map.putString(CVV_KEY, encryptedCard.encryptedSecurityCode)
      promise.resolve(map)
    } catch (e: EncryptionException) {
      promise.reject(ERROR_MESSAGE, e)
    }
  }

  @ReactMethod
  fun encryptBin(
    bin: String,
    publicKey: String,
    promise: Promise,
  ) {
    try {
      val encryptedBin = CardEncrypter.encryptBin(bin, publicKey)
      promise.resolve(encryptedBin)
    } catch (e: EncryptionException) {
      promise.reject(ERROR_MESSAGE, e)
    }
  }

  @ReactMethod
  fun validateCardNumber(
    cardNumber: String,
    enableLuhnCheck: Boolean,
    promise: Promise,
  ) {
    val validationResult = CardNumberValidator.validateCardNumber(cardNumber, enableLuhnCheck)
    promise.resolve(validationResult is CardNumberValidationResult.Valid)
  }

  @ReactMethod
  fun validateCardExpiryDate(
    expiryMonth: String,
    expiryYear: String,
    promise: Promise,
  ) {
    val expireMonth = expiryMonth.toIntOrNull()
    val expireYear = expiryYear.toIntOrNull()
    if (expireMonth == null || expireYear == null) {
      promise.resolve(false)
      return
    }

    val expiryDate = ExpiryDate(expireMonth, expireYear)
    val validationResult = CardExpiryDateValidator.validateExpiryDate(expiryDate)
    promise.resolve(validationResult is CardExpiryDateValidationResult.Valid)
  }

  @ReactMethod
  fun validateCardSecurityCode(
    securityCode: String,
    cardBrand: String?,
    promise: Promise,
  ) {
    val cardType = cardBrand?.let { CardBrand(it) }
    val validationResult = CardSecurityCodeValidator.validateSecurityCode(securityCode, cardType)
    promise.resolve(validationResult is CardSecurityCodeValidationResult.Valid)
  }

  companion object {
    private const val TAG = "AdyenCSE"
    private const val COMPONENT_NAME = "AdyenCSE"
    private const val NUMBER_KEY = "number"
    private const val EXPIRY_MONTH_KEY = "expiryMonth"
    private const val EXPIRY_YEAR_KEY = "expiryYear"
    private const val CVV_KEY = "cvv"
    private const val ERROR_MESSAGE = "Encryption failed"
  }
}

File: ios/CSE/AdyenCSEModule.swift
//
// Copyright (c) 2023 Adyen N.V.
//
// This file is open source and available under the MIT license. See the LICENSE file for more info.
//

import Adyen
import Foundation
import React

@objc(AdyenCSE)
internal final class AdyenCSEModule: NSObject {

    @objc
    static func requiresMainQueueSetup() -> Bool {
        true
    }

    @objc
    func encryptCard(_ payload: NSDictionary,
                     publicKey: NSString,
                     resolver: RCTPromiseResolveBlock,
                     rejecter: RCTPromiseRejectBlock) {
        do {
            let unencryptedCard = try Card(from: payload)
            let encryptedCard = try CardEncryptor.encrypt(card: unencryptedCard, with: publicKey as String)
            resolver(encryptedCard.jsonObject)
        } catch {
            rejecter(Constant.errorMessage, nil, error)
        }
    }

    @objc
    func encryptBin(_ bin: NSString,
                    publicKey: NSString,
                    resolver: RCTPromiseResolveBlock,
                    rejecter: RCTPromiseRejectBlock) {
        let formattedBin = bin.replacingOccurrences(of: " ", with: "")
        do {
            let encryptedBin = try CardEncryptor.encrypt(bin: formattedBin as String, with: publicKey as String)
            resolver(encryptedBin)
        } catch {
            rejecter("AdyenCSE", Constant.errorMessage, error)
        }
    }

    @objc
    func validateCardNumber(_ cardNumber: NSString,
                            enableLuhnCheck: Bool,
                            resolver: RCTPromiseResolveBlock,
                            rejecter _: RCTPromiseRejectBlock) {
        let isValid = CardNumberValidator(
            isLuhnCheckEnabled: enableLuhnCheck,
            isEnteredBrandSupported: true
        ).isValid(cardNumber as String)
        resolver(isValid)
    }

    @objc
    func validateCardExpiryDate(_ expiryMonth: NSString,
                                expiryYear: NSString,
                                resolver: RCTPromiseResolveBlock,
                                rejecter _: RCTPromiseRejectBlock) {
        let yearString = expiryYear as String
        let isValid: Bool
        if yearString.count == 2 || yearString.count == 4 {
            let lastTwoYearChars = String(yearString.suffix(2))
            isValid = CardExpiryDateValidator().isValid("\(expiryMonth)\(lastTwoYearChars)")
        } else {
            isValid = false
        }
        resolver(isValid)
    }

    @objc
    func validateCardSecurityCode(_ securityCode: NSString,
                                  cardBrand: NSString?,
                                  resolver: RCTPromiseResolveBlock,
                                  rejecter _: RCTPromiseRejectBlock) {
        if let cardBrand = cardBrand as String? {
            let brand = CardBrand(rawValue: cardBrand)
            resolver(CardSecurityCodeValidator(cardBrand: brand).isValid(securityCode as String))
        } else {
            resolver(CardSecurityCodeValidator().isValid(securityCode as String))
        }
    }

    private enum Constant {
        static var errorMessage = "Encryption failed"
    }
}
